chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,418 @@
+        state[0] = WORD_1;
+        state[1] = WORD_2;
+        state[2] = WORD_3;
+        state[3] = WORD_4;
+
+        // The next two rows are based on the session key.
+        for (state, word) in state[4..12].iter_mut().zip(key.0.chunks_exact(4)) {
+            *state =
+                u32::from_le_bytes(word.try_into().expect("chunks exact returns 4-byte slices"));
+        }
+
+        // The final row is the count and the nonce.
+        state[12] = count;
+
+        for (state, word) in state[13..16].iter_mut().zip(nonce.0.chunks_exact(4)) {
+            *state =
+                u32::from_le_bytes(word.try_into().expect("chunks exact returns 4-byte slices"));
+        }
+
+        State(state)
+    }
+
+    /// Each "quarter" round of ChaCha scrambles 4 of the 16 words (where the quarter comes from)
+    /// that make up the state using some Addition (mod 2^32), Rotation, and XOR (a.k.a. ARX).
+    fn quarter_round(&mut self, a: usize, b: usize, c: usize, d: usize) {
+        self.0[a] = self.0[a].wrapping_add(self.0[b]);
+        self.0[d] = (self.0[d] ^ self.0[a]).rotate_left(16);
+        self.0[c] = self.0[c].wrapping_add(self.0[d]);
+        self.0[b] = (self.0[b] ^ self.0[c]).rotate_left(12);
+        self.0[a] = self.0[a].wrapping_add(self.0[b]);
+        self.0[d] = (self.0[d] ^ self.0[a]).rotate_left(8);
+        self.0[c] = self.0[c].wrapping_add(self.0[d]);
+        self.0[b] = (self.0[b] ^ self.0[c]).rotate_left(7);
+    }
+
+    /// Transform the state by performing the ChaCha block function.
+    fn chacha_block(&mut self) {
+        let initial_state = self.0;
+        for _ in 0..10 {
+            for (a, b, c, d) in CHACHA_ROUND_INDICIES {
+                self.quarter_round(a, b, c, d);
+            }
+        }
+        for (modified, initial) in self.0.iter_mut().zip(initial_state.iter()) {
+            *modified = modified.wrapping_add(*initial)
+        }
+    }
+
+    /// Expose the 512-bit state as a byte stream.
+    fn keystream(&self) -> [u8; 64] {
+        let mut keystream: [u8; 64] = [0; 64];
+        for (k, s) in keystream.chunks_exact_mut(4).zip(self.0) {
+            k.copy_from_slice(&s.to_le_bytes());
+        }
+
+        keystream
+    }
+}
+
+/// The ChaCha20 stream cipher from RFC8439.
+///
+/// The 20-round IETF version uses a 96-bit nonce and 32-bit block counter. This is the
+/// variant used in the Bitcoin ecosystem including BIP324.
+#[derive(Debug)]
+pub struct ChaCha20 {
+    /// Secret session key shared by the parties communicating.
+    key: SessionKey,
+    /// A session key and nonce pair should only be used once.  
+    nonce: Nonce,
+    /// Internal block index of keystream.
+    block_count: u32,
+    /// Interal byte offset index of the block_count.
+    seek_offset_bytes: usize,
+}
+
+impl ChaCha20 {
+    /// Make a new instance of ChaCha20 from an index in the keystream.
+    pub fn new(key: SessionKey, nonce: Nonce, seek: u32) -> Self {
```

Comment:
In 74306fbde17bbbc09e44c388af56a081d8fb7dbf:

I'd suggest making every single `pub fn new*` method `const`. (There are some others that can be made `const` but not enough of them to be really useful, at least without doing ugly/inefficient things. But constructors are always useful to be `const`.)

Reply:
Makes sense to me, done.